### DISCUSSION TITLE
Record validations via JSON schemas

### DISCUSSION THREAD

**Discussion by EmilyGraceSeville7cf** [nushell/nushell#16639] [Ideas]
### Related problem

There are no suggestions about possible keys for `config` record.

### Describe the solution you'd like

I guess we can implement such nushell feature where users can associate records with JSON schemas to enable intellisence for records (suggest possible keys and their values). It should just validate static values, not those passed as variables. Personally, I can write schema for `config`. Associating schemas I guess can be done via comments:

```nushell
# schema=<path|url>
let-env config = {

}
```

### Additional context and details

When closure or block is expected, we can omit `type` key in JSON schema as there is no way to validate such complex thing in it. Note that regular intellisence should also work concurentely with such JSON schema one.

In shell this feature can work like tab-completion, where possible keys are suggested and their static values. When variable is put instead of static value in record it's type should be checked and when it's `any` no error should occur (as it may be possible the right one).

**Comment by amtoine** [comment]
> There are no suggestions about possible keys for config record.
>
i'm not sure i get this?
do you mean in the REPL?
in script? with or without LSP?
:smile:

**Comment by EmilyGraceSeville7cf** [comment]
I mean everywhere. :)

**Comment by amtoine** [comment]
so, for instance in the REPL, writing
```
let-env config = {color_c#<TAB>
```
would complete it with
```
let-env config = {color_config: #
```
> **Note**
> in the above snippets i denote the cursor position with a `#` :wink:

**Comment by EmilyGraceSeville7cf** [comment]
> so, for instance in the REPL, writing

Yep. ;)

**Comment by amtoine** [comment]
okey now, i get it, thanks :star_struck:

**Comment by EmilyGraceSeville7cf** [comment]
At the same time, maybe it's better to introduce some concept of classes like in other languages to enable such completion happen instead of JSON schema?

**Comment by 132ikl** [comment]
There's now completions for `$env.config`, although for cell path access and not while inside the actual record. I think the use case here is a bit outside of scope for Nushell, at least for pre-1.0. I'll move this to a discussion and we can reconsider this in the future if there's interest
